### PR TITLE
Top-levels FunctionDefinition to allow access to all function metadata

### LINKED ISSUE #637: Expose more information from CompiledModule
**Is your feature request related to a problem? Please describe.**
After having created a `CompiledModule`, I'd like to infer some information from it. For example, I'd like to know what type of functions it wants imported and what kind of functions it is exporting, amongst other things.

**Describe the solution you'd like**
`CompiledModule` (or actually, it's underlaying struct `compiledModule{}`) seems to embed `internal/wasm.Module{}` which appears to hold that information. Perhaps we could bubble up that information to the public API?

**Describe alternatives you've considered**
What I can do right now is create an instance from the `CompiledModule` and offer it all the imports available to me in the hope that it satisfies what the module needs. Then I can poke around with functions like `ExportedFunction()` and poll for information.

Obviously, this is not a great solution and this solution doesn't help me figure out what imports a module needs.

### PR DESCRIPTION
**Pull request by codefromthecrypt**
This top-levels `api.FunctionDefinition` which was formerly
experimental, and also adds import metadata to it. Now, it holds all
metadata known at compile time.

Here are the public API visible changes:
* api.ExportedFunction - replaced with api.FunctionDefinition as it is
  usable for all types of functions.
* api.Function - `.ParamTypes/ResultTypes()` are replaced with
  `.Definition().
* api.FunctionDefinition - extracted from experimental and adds
  `.Import()` to get the any imported module and function name.
* experimental.FunctionDefinition - replaced with
  api.FunctionDefinition.
* experimental.FunctionListenerFactory - adds first arg of the
  instantiated module name, as it can be different than compiled.
* wazero.CompiledModule - Adds `.ImportedFunctions()` and changes result
  type of `.ExportedFunctions()` to api.FunctionDefinition.

Internally, logic to create function definition are consolidated between
host and wasm-defined functions, notably wasm.Module now includes
`.BuildFunctionDefinitions()` which reduces duplication in
wasm.ModuleInstance `.BuildFunctions()`,

This obviates #681 by deleting the `ExportedFunction` type which
overlaps with this information.

This fixes #637 as it includes more metadata including imports.

### CODE REVIEW

File: internal/wasm/function_definition.go [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+	}
+	return
+}
+
+// BuildFunctionDefinitions generates function metadata that can be parsed from
+// the module. This must be called after all validation.
+//
+// Note: This is exported for tests who don't use wazero.Runtime or
+// NewHostModule to compile the module.
+func (m *Module) BuildFunctionDefinitions() *Module {
```
Comment (mathetake) [inline internal/wasm/function_definition.go:32]: do we need to return the self pointer?

**Review by mathetake — COMMENTED** [review]
now Is it possible to enhance the compilation error message with the actual name, not only the index here? https://github.com/tetratelabs/wazero/blob/c00afaa37e735049f360ab2595423d9410b7ee9c/internal/engine/compiler/engine.go#L444

File: internal/wasm/function_definition.go [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+func (f *functionDefinition) Name() string {
+	return f.name
+}
+
+// Import implements the same method as documented on api.FunctionDefinition.
+func (f *functionDefinition) Import() (moduleName, name string, isImport bool) {
+	if importDesc := f.importDesc; importDesc != nil {
+		return importDesc[0], importDesc[1], true
+	}
+	return "", "", false
```
Comment (mathetake) [inline internal/wasm/function_definition.go:130]: less 
```suggestion
	if importDesc := f.importDesc; importDesc != nil {
		moduleName, name, isImport = importDesc[0], importDesc[1], true
	}
	return
```

**Review by mathetake — APPROVED** [review]
much better thanks!

File: internal/wasm/function_definition.go (hunk 1 above)
Comment (codefromthecrypt) [reply internal/wasm/function_definition.go:32]: no, just did it for convenience when inlining wasm modules

**Comment by codefromthecrypt** [comment]
> now Is it possible to enhance the compilation error message with the actual name, not only the index here? 

ok now we use debugname everywhere

File: internal/engine/compiler/compiler_test.go [hunk 3]
```diff
@@ -19,6 +20,104 @@ func TestMain(m *testing.M) {
 	os.Exit(m.Run())
 }
 
+// Ensures that the offset consts do not drift when we manipulate the target
+// structs.
+//
+// Note: This is a package initializer as many tests could fail if these
+// constants are misaligned, hiding the root cause.
+func init() {
```
Comment (codefromthecrypt) [inline internal/engine/compiler/compiler_test.go:28]: this is new and allows an early break on offset problem

**Comment by codefromthecrypt** [comment]
thanks for the help @mathetake!

**Pull request merged**